controllers/infra/drivers/aliyun/delete_instance.go

```diff
@@ -54,86 +56,145 @@ func (d Driver) deleteInstances(hosts *v1.Hosts) error {
 		return nil
 	}
 
+	// detach keypair
+	//detachKeyPairRequest := &ecs.DetachKeyPairRequest{
+	//	RpcRequest:  ecs.CreateDetachKeyPairRequest().RpcRequest,
+	//	InstanceIds: "[\"" + strings.Join(instanceIDs, ",") + "\"]",
+	//	KeyPairName: instance.KeyPairName,
+	//}
+
+	//logger.Info("Detaching keypair %s from instances %v", instance.KeyPairName, instanceIDs)
+	//_, err := DetachKeyPair(client, detachKeyPairRequest)
+	//if err != nil {
+	//	return fmt.Errorf("got an error detaching keypair %s from instances %v: %v", instance.KeyPairName, instanceIDs, err)
+	//}
+
+	// delete instances
+	deleteInstancesRequest := &ecs.DeleteInstancesRequest{
+		RpcRequest: ecs.CreateDeleteInstancesRequest().RpcRequest,
+		InstanceId: &instanceIDs,
+		Force:      "true",
+	}
+	logger.Info("Deleting Aliyun ECS instances: %v", instanceIDs)
+	err := retry.Retry(30, 10*time.Second, func() error {
+		_, err := DeleteInstances(client, deleteInstancesRequest)
+		if err != nil {
+			return fmt.Errorf("got an error deleting Aliyun ECS instances: %v", err)
+		}
+		return nil
+	})
+	if err != nil {
+		return fmt.Errorf("delete instances failed: %v", err)
+	}
+	return nil
+}
+
+func (d Driver) deleteInfra(infra *v1.Infra) error {
+	if len(infra.Spec.Hosts) == 0 {
+		return nil
+	}
+	client := d.Client
+	eg, _ := errgroup.WithContext(context.Background())
+	var instanceID string
+	for _, hosts := range infra.Spec.Hosts {
+		if len(hosts.Metadata) != 0 {
+			instanceID = hosts.Metadata[0].ID
+		}
+	}
 	// get instance info
 	describeInstanceRequest := &ecs.DescribeInstancesRequest{
 		RpcRequest:  ecs.CreateDescribeInstancesRequest().RpcRequest,
-		InstanceIds: "[\"" + instanceIDs[0] + "\"]",
+		InstanceIds: "[\"" + instanceID + "\"]",
 	}
 	describeInstanceResponse, err := GetInstances(client, describeInstanceRequest)
 	if err != nil || len(describeInstanceResponse.Instances.Instance) == 0 {
 		return fmt.Errorf("got an error getting instance info: %v", err)
 	}
-
-	// detach keypair
 	instance := describeInstanceResponse.Instances.Instance[0]
-	detachKeyPairRequest := &ecs.DetachKeyPairRequest{
-		RpcRequest:  ecs.CreateDetachKeyPairRequest().RpcRequest,
-		InstanceIds: "[\"" + strings.Join(instanceIDs, ",") + "\"]",
-		KeyPairName: instance.KeyPairName,
+
+	// delete instances
+	for i := range infra.Spec.Hosts {
+		hosts := infra.Spec.Hosts[i]
+		eg.Go(func() error {
+			return d.deleteInstances(&hosts)
+		})
+	}
+	if err := eg.Wait(); err != nil {
+		return fmt.Errorf("got an error deleting instances: %v", err)
```

Comment:
failed to ...

Reply:
> failed to ...

okay I've changed it (this log is generated by copilot, lol